lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,369 @@
+        # dimension then place within list.
+        try:
+            slices_over_realization = temperature_cube.slices_over(
+                "realization")
+        except iris.exceptions.CoordinateNotFoundError:
+            slices_over_realization = [temperature_cube]
+
+        # Creates cube list to hold lapse rate data.
+        lapse_rate_cube_list = iris.cube.CubeList([])
+
+        for temp_slice in slices_over_realization:
+
+            # Create slice to store lapse rate values.
+            lapse_rate_slice = temp_slice
+
+            temperature_data = temp_slice.data
+
+            # Fill sea points with NaN values. Can't use Numpy mask since not
+            # recognised by "generic_filter" function.
+            temperature_data = np.where(land_sea_mask, temperature_data,
+                                        np.nan)
+
+            # Saves all neighbourhoods into "all_temp_subsections".
+            # cval is value given to points outside the array.
+            fnc = SaveNeighbourhood(allbuffers=all_temp_subsections)
+            generic_filter(temperature_data, fnc.filter,
+                           size=self.nbhood_size,
+                           mode='constant', cval=np.nan)
+
+            fnc = SaveNeighbourhood(allbuffers=all_orog_subsections)
+            generic_filter(orography_data, fnc.filter,
+                           size=self.nbhood_size,
+                           mode='constant', cval=np.nan)
+
+            # height_diff_mask is True for points where the height
+            # difference between the central point and its neighbours
+            # is > max_height_diff.
+            height_diff_mask = self._create_heightdiff_mask(
+                all_orog_subsections)
+
+            # Mask points with extreme height differences as NaN.
+            all_orog_subsections = np.where(height_diff_mask, np.nan,
+                                            all_orog_subsections)
+            all_temp_subsections = np.where(height_diff_mask, np.nan,
+                                            all_temp_subsections)
+
+            # Loop through both arrays and find gradient of each subsection.
+            # The gradient indicates lapse rate - save into another array.
+            # TODO: This for loop is the bottleneck in the code - 1 array takes
+            # approximately 29 seconds out of a total running time of 33
+            # seconds for UKVX data. This section needs to be parallelised.
+            lapse_rate_array = [self._calc_lapse_rate(temp, orog)
+                                for temp, orog in zip(all_temp_subsections,
+                                                      all_orog_subsections)]
+
+            lapse_rate_array = np.array(
+                lapse_rate_array, dtype=np.float32).reshape(dataarray_shape)
+
+            # According to the paper - any value below DALR is reset to DALR.
+            # HOWEVER the FORTRAN code also imposes an upper limit (see class
```

Comment:
This comment doesn't belong here. You have implemented an upper and lower bound. You can specify this, but we should not reference random Fortran code. Having an upper and lower bound is perfectly reasonable, and I don't think requires any justification here. What these bounds might be is one of the many things to explore when time allows.